File: src/Study/Bad/12865.cpp
#include <iostream>

#define endl "\n"
#define MAX 110
#define Max(A,B) (((A)>(B)) ? (A) : (B))

using namespace std;

int n,k;
int Weight[MAX];
int Value[MAX];
int DP[MAX][100010];


void Input(){
	cin >> n >> k;
	for(int i = 1; i <= n; ++i){
		cin >> Weight[i] >> Value[i];
	}
}

void Solution(){
	for(int i = 1; i <= n; ++i){
		for(int j = 1; j <= k; ++j){
			if (j >= Weight[i])    
				DP[i][j] = Max(DP[i - 1][j], DP[i - 1][j - Weight[i]] + Value[i]);
            else 
				DP[i][j] = DP[i - 1][j];
		}
	}
	cout << DP[n][k] << endl;
}

void Solve(){
	Input();
	Solution();
}

int main(){
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	Solve();
	
	return 0;
}

File: src/Study_Python/Solved/12865.py
n, k = map(int, input().split())

arr = [[0,0]]
knapsack = [[0 for _ in range(k+1)] for _ in range(n+1)]


for _ in range(n):
    arr.append(list(map(int,input().split())))

#print(knapsack)
#print(arr)

for i in range(1, n + 1):
    for j in range(1, k + 1):
        weight = arr[i][0]
        value = arr[i][1]
        #print(str(weight) + " " + str(value))

        if j < weight:
            knapsack[i][j] = knapsack[i - 1][j]
        else:
            knapsack[i][j] = max(value + knapsack[i - 1][j - weight], knapsack[i - 1][j])

print(knapsack[n][k])
